src/audio/drc/drc.c

```diff
@@ -0,0 +1,605 @@
+
+#include <sof/audio/component.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/drc/drc.h>
+#include <sof/audio/drc/drc_math.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/math/decibels.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/eq.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_drc;
+
+/* TODO: get the real UUID */
+/* 5150c0e6-27f9-4ec8-8351-c705b642d12f */
+DECLARE_SOF_RT_UUID("drc", drc_uuid, 0x5150c0e6, 0x27f9, 0x4ec8,
+		    0x83, 0x51, 0xc7, 0x05, 0xb6, 0x42, 0xd1, 0x2f);
+
+DECLARE_TR_CTX(drc_tr, SOF_UUID(drc_uuid), LOG_LEVEL_INFO);
+
+static inline void drc_free_parameters(struct sof_drc_config **config)
+{
+	rfree(*config);
+	*config = NULL;
+}
+
+static inline void drc_reset_state(struct drc_state *state)
+{
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; ++i) {
+		rfree(state->pre_delay_buffers[i]);
+		state->pre_delay_buffers[i] = NULL;
+	}
+
+	state->detector_average = 0;
+	state->compressor_gain = Q_CONVERT_FLOAT(1.0f, 30);
+
+	state->last_pre_delay_frames = DRC_DEFAULT_PRE_DELAY_FRAMES;
+	state->pre_delay_read_index = 0;
+	state->pre_delay_write_index = DRC_DEFAULT_PRE_DELAY_FRAMES;
+
+	state->envelope_rate = 0;
+	state->scaled_desired_gain = 0;
+
+	state->processed = 0;
+
+	state->max_attack_compression_diff_db = INT32_MIN;
+}
+
+static inline int drc_init_pre_delay_buffers(struct drc_state *state,
+					     size_t sample_bytes,
+					     int channels)
+{
+	int i;
+
+	/* Allocate pre-delay (lookahead) buffers */
+	for (i = 0; i < channels; ++i) {
+		state->pre_delay_buffers[i] =
+			rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				sample_bytes * DRC_MAX_PRE_DELAY_FRAMES);
+		if (!state->pre_delay_buffers[i])
+			return -ENOMEM;
+		memset(state->pre_delay_buffers[i], 0, sample_bytes * DRC_MAX_PRE_DELAY_FRAMES);
+	}
+
+	return 0;
+}
+
+static inline void drc_set_pre_delay_time(struct drc_state *state,
+					  int32_t pre_delay_time,
+					  int32_t rate)
+{
+	uint32_t pre_delay_frames;
+
+	/* Re-configure look-ahead section pre-delay if delay time has
+	 * changed. */
+	pre_delay_frames = pre_delay_time * rate;
+	pre_delay_frames = min(pre_delay_frames, DRC_MAX_PRE_DELAY_FRAMES - 1);
+
+	/* Make pre_delay_frames multiplies of DIVISION_FRAMES. This way we
+	 * won't split a division of samples into two blocks of memory, so it is
+	 * easier to process. This may make the actual delay time slightly less
+	 * than the specified value, but the difference is less than 1ms. */
+	pre_delay_frames &= ~DRC_DIVISION_FRAMES_MASK;
```

Comment:
It would be good to add to drc.h a note that DRC_DIVISION_FRAMES needs to be a 2^N number for this to work correctly.